From 68249d6fef8ddb474dc5604f6bd3afb1d2a27539 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Thu, 9 Oct 2014 21:00:41 -0400
Subject: [fs] xfs: only set extent size hint when asked

Message-id: <1412888441-4181-48-git-send-email-sandeen@redhat.com>
Patchwork-id: 97327
O-Subject: [RHEL7.1 PATCH 47/47] [fs] xfs: only set extent size hint when asked
Bugzilla: 1145837
RH-Acked-by: Brian Foster <bfoster@redhat.com>
RH-Acked-by: Dave Chinner <dchinner@redhat.com>

Bugzilla: 1145837
Upstream Status: Reviewed on list upstream
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=8082426
Tested: locally using xfstests

    xfs: only set extent size hint when asked

    From: Dave Chinner <dchinner@redhat.com>

    Currently the extent size hint is set unconditionally in
    xfs_ioctl_setattr(), even when the FSX_EXTSIZE flag is not set. This
    means we can set values from uninitialised stack variables. Hence
    only set the extent size hint from userspace when both the mask
    falg is set and the inode has the XFS_DIFLAG_EXTSIZE flag set to
    indicate that we should have an extent size hint set on the inode.

    Signed-off-by: Dave Chinner <dchinner@redhat.com>
    Reviewed-by: Brian Foster <bfoster@redhat.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/xfs/xfs_ioctl.c b/fs/xfs/xfs_ioctl.c
index a58bab5..c01d09b 100644
--- a/fs/xfs/xfs_ioctl.c
+++ b/fs/xfs/xfs_ioctl.c
@@ -1267,13 +1267,25 @@ xfs_ioctl_setattr(
 
  }
 
- if (mask & FSX_EXTSIZE)
-  ip->i_d.di_extsize = fa->fsx_extsize >> mp->m_sb.sb_blocklog;
  if (mask & FSX_XFLAGS) {
   xfs_set_diflags(ip, fa->fsx_xflags);
   xfs_diflags_to_linux(ip);
  }
 
+ /*
+  * Only set the extent size hint if we've already determined that the
+  * extent size hint should be set on the inode. If no extent size flags
+  * are set on the inode then unconditionally clear the extent size hint.
+  */
+ if (mask & FSX_EXTSIZE) {
+  int extsize = 0;
+
+  if (ip->i_d.di_flags &
+    (XFS_DIFLAG_EXTSIZE | XFS_DIFLAG_EXTSZINHERIT))
+   extsize = fa->fsx_extsize >> mp->m_sb.sb_blocklog;
+  ip->i_d.di_extsize = extsize;
+ }
+
  xfs_trans_ichgtime(tp, ip, XFS_ICHGTIME_CHG);
  xfs_trans_log_inode(tp, ip, XFS_ILOG_CORE);
 
-- 
1.7.1